File: code/algo/getoffer/go/2/find_bst_level_maxval.go
package main

func largestValues(root *TreeNode) (ans []int) {
	var dfs func(*TreeNode, int)
	dfs = func(node *TreeNode, curHeight int) {
		if node == nil {
			return
		}
		if curHeight == len(ans) {
			ans = append(ans, node.Val)
		} else {
			ans[curHeight] = max(ans[curHeight], node.Val)
		}
		dfs(node.Left, curHeight+1)
		dfs(node.Right, curHeight+1)
	}
	dfs(root, 0)
	return
}

File: code/algo/getoffer/c++/2/find_bst_level_maxval.cpp
#include <algorithm>
#include <vector>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
  std::vector<int> largestValues(TreeNode *root) {
    std::vector<int> ans;
    dfs(root, 0, ans);
    return ans;
  }

private:
  void dfs(TreeNode *node, int curHeight, std::vector<int> &ans) {
    if (!node) {
      return;
    }
    if (curHeight == ans.size()) {
      ans.push_back(node->val);
    } else {
      ans[curHeight] = std::max(ans[curHeight], node->val);
    }
    dfs(node->left, curHeight + 1, ans);
    dfs(node->right, curHeight + 1, ans);
  }
};

int max(int a, int b) { return b > a ? b : a; }
